### PR TITLE
 Close join probe before creating outer position iterator 

### LINKED ISSUE #9766: VerifyException exception when executing RIGHT OUTER JOIN
Hi,

I use presto 0.192 + HDP 2.5.3 + hive connector.

VerifyException occured in the following query.
```
SELECT yyyymmdd FROM test_view WHERE yyyymmdd='20180117' LIMIT 1
```

view ddl
```
create or replace view test_view
as
select
    b.yyyymm
    , b.yyyymmdd
...
    , count(distinct ...) as ... 
from
    ... a
right outer join
    (select distinct yyyymmdd, substr(yyyymmdd,1,6) as yyyymm, ... from ... cross join ...) b
on
    a.yyyymmdd = b.yyyymmdd
    and ...
group by
    b.yyyymm
    , b.yyyymmdd
...
```

stacktrace
```
com.google.common.base.VerifyException: undefined
	at com.google.common.base.Verify.verify(Verify.java:99)
	at com.facebook.presto.operator.PartitionedLookupSource$OuterPositionTracker$Factory.getOuterPositionIterator(PartitionedLookupSource.java:299)
	at com.facebook.presto.operator.PartitionedLookupSource$1.getOuterPositionIterator(PartitionedLookupSource.java:65)
	at com.facebook.presto.operator.PartitionedLookupSourceFactory.getOuterPositionIterator(PartitionedLookupSourceFactory.java:362)
	at com.facebook.presto.operator.LookupJoinOperatorFactory$PerLifespanData.lambda$new$1(LookupJoinOperatorFactory.java:315)
	at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:239)
	at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:229)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:902)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:813)
	at com.google.common.util.concurrent.AbstractFuture.setFuture(AbstractFuture.java:713)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.setResult(AbstractTransformFuture.java:221)
	at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.setResult(AbstractTransformFuture.java:200)
	at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:177)
	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:902)
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:813)
	at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:655)
	at com.google.common.util.concurrent.SettableFuture.set(SettableFuture.java:48)
	at com.facebook.presto.operator.ReferenceCount.release(ReferenceCount.java:55)
	at com.google.common.io.Closer.close(Closer.java:216)
	at com.facebook.presto.operator.LookupJoinOperator.close(LookupJoinOperator.java:514)
	at com.facebook.presto.operator.Driver.destroyIfNecessary(Driver.java:496)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:667)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:272)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at com.facebook.presto.execution.executor.LegacyPrioritizedSplitRunner.process(LegacyPrioritizedSplitRunner.java:23)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

If I modify view not to use ```right outer join``` in the following, error will be fixed although I'm not sure why
```
create or replace view test_view
as
select
    a.yyyymm
    , a.yyyymmdd
...
    , count(distinct ...) as ... 
from
    (select distinct yyyymmdd, substr(yyyymmdd,1,6) as yyyymm, ... from ... cross join ...) a
left outer join
  ... b
on
    a.yyyymmdd = b.yyyymmdd
    and ...
group by
    a.yyyymm
    , a.yyyymmdd
...
```

Thanks

### PR DESCRIPTION
**Pull request by haozhun**
Fixes #9766.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 1]
```diff
@@ -507,10 +507,13 @@ public void close()
         probe = null;
 
         try (Closer closer = Closer.create()) {
+            // `afterClose` must be run last.
+            // Closer is documented to mimic try-with-resource, which implies close will happen in reverse order.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: `// Closer closes the resources in the reverse order` would be enough

**Comment by haozhun** [comment]
An outer join query fails consistently with this stacktrace during our release verification process. I'm still trying to find out what changed since the last release that could have lead to this.

```
Query 20180216_205804_00001_z2u8m failed: null                                                                                                                                                                         [46/1829]com.google.common.base.VerifyException
        at com.google.common.base.Verify.verify(Verify.java:99)
        at com.facebook.presto.operator.PartitionedLookupSource$OuterPositionTracker$Factory.getOuterPositionIterator(PartitionedLookupSource.java:299)
        at com.facebook.presto.operator.PartitionedLookupSource$1.getOuterPositionIterator(PartitionedLookupSource.java:65)
        at com.facebook.presto.operator.PartitionedLookupSourceFactory.getOuterPositionIterator(PartitionedLookupSourceFactory.java:362)
        at com.facebook.presto.operator.LookupJoinOperatorFactory$PerLifespanData.lambda$new$1(LookupJoinOperatorFactory.java:315)
        at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:239)
        at com.google.common.util.concurrent.AbstractTransformFuture$TransformFuture.doTransform(AbstractTransformFuture.java:229)
        at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:130)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:902)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:813)
        at com.google.common.util.concurrent.AbstractFuture.setFuture(AbstractFuture.java:713)
        at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.setResult(AbstractTransformFuture.java:221)
        at com.google.common.util.concurrent.AbstractTransformFuture$AsyncTransformFuture.setResult(AbstractTransformFuture.java:200)
        at com.google.common.util.concurrent.AbstractTransformFuture.run(AbstractTransformFuture.java:177)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:399)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:902)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:813)
        at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:655)
        at com.google.common.util.concurrent.SettableFuture.set(SettableFuture.java:48)
        at com.facebook.presto.operator.ReferenceCount.release(ReferenceCount.java:55)
        at com.google.common.io.Closer.close(Closer.java:216)
        at com.facebook.presto.operator.LookupJoinOperator.close(LookupJoinOperator.java:514)
        at com.facebook.presto.operator.Driver.destroyIfNecessary(Driver.java:496)
        at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:653)
        at com.facebook.presto.operator.Driver.processFor(Driver.java:272)
        at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:975)
        at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
        at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:492)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

**Comment by findepi** [comment]
This is the same stacktrace as in https://github.com/prestodb/presto/issues/9766, right? FWIW, that issue was reported against 0.192.

**Comment by haozhun** [comment]
@findepi This is very helpful finding.

**Comment by haozhun** [comment]
@findepi Ah, I forgot to address you comments.

**Pull request merged**